### ISSUE TITLE
Wrong file and sources properties are generated on external sourcefiles

### ISSUE THREAD

**Issue by Emigre** [gulp-sourcemaps/gulp-sourcemaps#163]
When writing external sourcemaps, the `file` and `path` properties are currently set incorrectly in the sourcemap. 

According to the spec, all non-absolute URLs inside the sourcemap have to be relative to the location of the sourcemap itself (see https://github.com/mishoo/UglifyJS2/issues/700).

Currently, in this example...

``` javascript
gulp.src('app/src/**/*.less')
    .pipe(gulp.dest('.tmp/src/'))
    .pipe(sourcemaps.init())
      .pipe(less())
    .pipe(sourcemaps.write('.', { includeContent: false }))
    .pipe(gulp.dest('.tmp/src/'));
```

...for a file located in `app/src/components/main/main.less`, you would expect the following end result inside the `.tmp` folder (note that we copy the less file in the second line):

```
.tmp/
  └ src/
     └ components/
           └ main/
               ├ main.css
               ├ main.css.map
               └ main.less
```

Which should have `"sources": ["main.less"]` and `"file":"main.css"` set in that sourcemap file. Instead, the paths in those properties are set relative to the path passed to write as its first argument (in this case `'.'`) that is `"sources": ["components/main/main.less"]` and `"file":"components/main/main.css"`.

There are currently two pull requests by @sixinli that try to fix this issue. One is floridoo/gulp-sourcemaps/pull/157 which fixes how the `file` property is set. The other is floridoo/gulp-sourcemaps/pull/156 which does the same for the `sources` property.

In the second pull request there is the complication that, to be able to calculate the **relative** path to the original source file, we need to know the destination folder. In the example shown the less file is in the same folder than the sourcemap, but if I had not copied it, we would need to know that the destination folder is `.tmp/src/` (set in the last line) at the time of calling `.pipe(sourcemaps.write('.', { includeContent: false }))`, to be able to construct the necessary path `../../../../ etc.`. 

As we only know that path at the end of the pipe, @sixinli's fix includes an option that you have to the pugin to set what's the final path, like this:

``` javascript
      .pipe(sourcemaps.write('.', {
        includeContent: false,
        relativeToSourcePath: 'dist/build'
      }))
```

There is a small error that I have noticed in the second pull request about how the path is formed. I'm going to send @sixinli a patch for it, but apart from that both of them solve the issue.

**Comment by Emigre** [comment]
While this is not fixed I am using a branch in a forked repo I've done.

```javascript
  "gulp-sourcemaps": "Emigre/gulp-sourcemaps#relative_paths",
```

I thought about letting you know @sixinli. This branch applies both of your merge requests.

**Comment by sixinli** [comment]
good to know :)

**Comment by meriturva** [comment]
@Emigre Great! You have saved my day!
So thanks!

**Comment by meriturva** [comment]
@floridoo have you seen that issue?

**Comment by Emigre** [comment]
cool, man : )

**Comment by meriturva** [comment]
I think that this project in not anymore maintained......
What do you think @Emigre and @floridoo ?

**Comment by shinnn** [comment]
https://github.com/floridoo/gulp-sourcemaps/issues/94#issuecomment-165164311

**Comment by Emigre** [comment]
I guess... I imagine that he probably doesn't have the time at the moment to invest in this.

**Comment by amcdnl** [comment]
@floridoo - can we get this merged?? without this, I had to write this nasty code:

```
gulp.task('babel', function () {
  return gulp.src('./src/**/*.js', { base: './src' })
    .pipe(sourcemaps.init())
    .pipe(babel())
    .pipe(sourcemaps.write({
      includeContent: false,
      debug: true,
      sourceRoot: function(file){
        var from = file.path;
        var to = path.resolve(path.join(__dirname, '..', 'src'));
        var dest = path.join(path.relative(from, to), 'src');
        if (isWin){
          dest = dest.replace(/\\/g, '/');
        }
        return dest;
      }
    }))
    .pipe(gulp.dest('./dist'));
});
```

**Comment by Emigre** [comment]
Perhaps this is fixed in the upcoming version 2.0.0

I haven't had the time to try the alpha

**Comment by meriturva** [comment]
so @Emigre .. will be a 2.0 version?

**Comment by Emigre** [comment]
It seems so. If you check the last commit in the master branch, it has been tagged as v2.0.0-alpha last April. It has not been published to the npm registry yet, though.

**Comment by nmccready** [comment]
@Emigre did you ever confirm that this is resolved in 2.X? If not can you PR your work? I am trying to help @floridoo get caught up and close issues.

Also does anyone know the original breaking changes that caused 2.X?

**Comment by nmccready** [comment]
@Emigre I'll look into merging those two original PR's manually

**Comment by Emigre** [comment]
Hey @nmccready, sorry for the delay in getting back to you, I've been quite busy lately. Mmmm, I haven't taken a look at this in a while... I don't know if it has been resolved in 2.X. Have you found it out? Cheers.

**Comment by nmccready** [comment]
I looked across the code against the two fixes; anyway I don't think either of them are in the current code. So my guess is they are still broken.

**Comment by Emigre** [comment]
I see! I'm happy to give a hand if I'm able, it'll have to be next week or so, though. I won't have a lot of time in the next few days.

**Comment by nmccready** [comment]
@Emigre cool yes any help is welcome; also I would add more admins to the project but I don't have the perms. But I can request it if you remain active.

If your comparing code against the latest; I moved files around to make it more modular and easier to read. (to me anyways)

**Comment by phated** [comment]
@Emigre I'm looking into this in relation to my work on vinyl-sourcemap.  It seems that the `file` property has been fixed but not the `sources` array.  I'm thinking that if this can be completely fixed, the `mapSources` and `mapFile` options can be removed (and maybe even `options.destPath`).

**Comment by phated** [comment]
@Emigre another thing I found while digging at https://github.com/gulpjs/vinyl-sourcemap/issues/18

**Comment by phated** [comment]
And my solution to fix the `sources` property is something like (in `mapDestPath`):

```js
sourceMap.sources = sourceMap.sources.map(function(sourcePath) {
    var absSource = path.join(file.base, sourcePath);
    return path.relative(path.dirname(sourceMapPath), absSource);
});
```

Since the re-mapping of sources happens after the files are read, it doesn't affect the `includeContents` option.

**Comment by meriturva** [comment]
@phated you are right.....still problem with sources and i still use version from @Emigre due of that.
Any news here?

**Comment by Emigre** [comment]
@phated, @nmccready, I'm sorry, I have thought about taking a look at this now and then, but I'm afraid that I just don't have the time... I'm quite busy... I hope this is going well. Cheers!

**Comment by gjdass** [comment]
+1.
Also having troubles with that.

Directory structure :
```text
src
|   App.ts
|   routes/
    |    Users.router.ts
dist
|   App.js
|   App.map.js
|   routes/
    |    Users.router.js
    |    Users.router.map.js
```

I wanted inline sources (at first). But seeing it was not working, I put the `write('.')` first argument in order to produce external map files and debug `[source]` values.

```javascript
.pipe(sourcemaps.write('.'))
```
produces :

```json
{"version":3,"sources":["../src/App.ts"],"file":"App.js"}
```
```json
{"version":3,"sources":["../src/routes/Users.router.ts"],"file":"Users.router.js"}
```

For the App.js it's fine. VSCode sees my breakpoints into App.ts, but **not into Users.router.ts**, and it seems logic regarding the `sources` value for that map file missing a relative `../`.
Adding it manually is working.

I assume it's the same problem for the `inline` sourcemaps ?
I'm not an expert and I had to spend a lot of time to understand how the plugin works to be honest so be nice 😮

**Comment by Emigre** [comment]
@nmccready I have tried version 2.6.0 and it seems to fix this issue, so I'm going to close this. 

To recapitulate, now in 2.6.0 instead of using the `relativeToSourcePath` option you are supposed to use the `destPath` option.  In the project that I've used to test this, this option creates the following map file:
```
{"version":3,"sources":["components/main/main-controller.less"],"names":[],"mappings":"AAEA;EACE,gBAAA;;AADF,cAEE;EACE,cAAA;EACA,kBAAA;;AAJJ,cAME,KAAI;EACF,eAAA;EACA,cAAA","file":"main-controller.css","sourceRoot":"../.."}
```
In this particular case the less file has been built in `components/main`, next to the map file and the css file. The `sourceRoot` option now points to the right location (`../..`) for the path to the source (`components/main/`) to reference the original file `main-controller.less`, which is what is expected and solves the bug.
I'm not sure who has fixed the issue but thank you so much for your contribution and the effort put in the plugin. 👍  :)
